### PR TITLE
Allow to sort dynamic fields

### PR DESCRIPTION
**Pull request by Victorcorcos**
# Description ✍️

Allow to sort dynamic fields with this query:

* `order: { dynamicFields: { integer_test: ASC, string_test: DESC } }`

→ Graphield Declaration

```rb
class MyClass
  include Mongoid::Document
  include Mongoid::Attributes::Dynamic
  include Graphoid::Queries
  include Graphoid::Mutations
  include Graphoid::Graphield

  [...]

  graphield :dynamic_fields, Hash, dynamic: true

  [...]
end
```

→ GraphQL Query Example

```rb
{
    records(where: { sheetId: "62c46be8a6af030113ffb9bf" }, order: { dynamicFields: { integer_test: ASC, string_test: DESC } }) {
        count
        data {
            id
            dynamicFields
        }
    }
}
```

The possibility to sort records by dynamic fields is allowed by the changes of this PR plus this new code on the `__sort_option__` (inside `mongoid` gem):

* `gems/mongoid-7.0.13/lib/mongoid/criteria/queryable/extensions/hash.rb`

Screen Shot 2022-12-01 at 15 36 57

# Overview 🔍

Screen Shot 2022-11-29 at 00 17 37

# Checks ☑️

- [x] Allow to sort dynamic fields
- [ ] Smoke Tests

### CODE REVIEW

File: lib/graphoid/definitions/sorter.rb [hunk 1]
```diff
@@ -13,6 +14,7 @@ def generate(model)
 
           Attribute.fields_of(model).each do |field|
             name = Utils.camelize(field.name)
+            next argument(name, Sorter.dynamic_type) if name == 'dynamicFields'
```
Comment (niltonvasques) [inline lib/graphoid/definitions/sorter.rb:17]: What about instead have our field here hardcoded, we have some kind of metaprogramming informing that the field is dynamic? Find an idea

image

**Review by niltonvasques — CHANGES_REQUESTED** [review]
:diving_mask:

File: lib/graphoid/definitions/sorter.rb (hunk 1 above)
Comment (Victorcorcos) [reply lib/graphoid/definitions/sorter.rb:17]: Great idea! 🧠 

Let me see how to do it

**Comment by Victorcorcos** [comment]
Done @niltonvasques

File: lib/graphoid/operators/attribute.rb [hunk 2]
```diff
@@ -2,14 +2,15 @@
...
+    attr_reader :name, :type, :opts
 
     PERMS = %i[read create update delete]
 
-    def initialize(name:, type:)
+    def initialize(name:, type:, opts: {})
       @name = name.to_s
       @camel_name = Utils.camelize(@name)
       @type = type
+      @opts = opts
```
Comment (niltonvasques) [inline lib/graphoid/operators/attribute.rb:13]:
```suggestion
      @opts = opts
    end    
   
    def dynamic?
      @opts[:dynamic] == true
    end
```

**Review by niltonvasques — CHANGES_REQUESTED** [review]
By doing this we can avoid the try dig on the other file

File: lib/graphoid/definitions/sorter.rb [hunk 3]
```diff
@@ -13,6 +14,7 @@ def generate(model)
 
           Attribute.fields_of(model).each do |field|
             name = Utils.camelize(field.name)
+            next argument(name, Sorter.dynamic_type) if field.try(:opts)&.dig(:dynamic)
```
Comment (Victorcorcos) [inline lib/graphoid/definitions/sorter.rb:17]:
```suggestion
            next argument(name, Sorter.dynamic_type) if field.dynamic?
```

**Review by niltonvasques — CHANGES_REQUESTED** [review]
Is working ? 

image


- [ ] missing tests

**Comment by Victorcorcos** [comment]
> Is working ?

Yes, the Digitalize PR is pointing to this PR and the tests are passing (the last automated tests is executing this version).

File: lib/graphoid/operators/attribute.rb [hunk 4]
```diff
@@ -2,14 +2,20 @@
...
+    end    
+   
+    def dynamic?
+      @opts[:dynamic] == true
+    end
```
Comment (Victorcorcos) [inline lib/graphoid/operators/attribute.rb:18]:
```suggestion
```

**Review by niltonvasques — APPROVED** [review]

**Pull request merged**